features/mbedtls/targets/TARGET_NXP/aes_alt.c

```diff
@@ -0,0 +1,609 @@
+/*
+ * Copyright 2020 NXP
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+#include "mbedtls/aes.h"
+#include "fsl_common.h"
+
+/******************************************************************************/
+/**************************** DCP *********************************************/
+/******************************************************************************/
+#if defined(FSL_FEATURE_SOC_DCP_COUNT) && (FSL_FEATURE_SOC_DCP_COUNT > 0)
+static dcp_handle_t s_dcpHandle = {.channel = kDCP_Channel0, .keySlot = kDCP_KeySlot0, .swapConfig = kDCP_NoSwap};
+#endif
+
+/******************************************************************************/
+/************************* Key slot management ********************************/
+/******************************************************************************/
+#if (defined(FSL_FEATURE_SOC_CAU3_COUNT) && (FSL_FEATURE_SOC_CAU3_COUNT > 0)) || (defined(MBEDTLS_FREESCALE_DCP_AES))
+static const void *s_mbedtlsCtx[4] = {0};
+
+static void crypto_attach_ctx_to_key_slot(const void *ctx, uint8_t keySlot)
+{
+    s_mbedtlsCtx[keySlot] = ctx;
+}
+
+static void crypto_detach_ctx_from_key_slot(const void *ctx)
+{
+    for (int i = 0; i < 4; i++)
+    {
+        if (ctx == s_mbedtlsCtx[i])
+        {
+            s_mbedtlsCtx[i] = NULL;
+            break;
+        }
+    }
+}
+
+static bool crypto_key_is_loaded(const void *ctx)
+{
+    bool ret = false;
+    for (int i = 0; i < 4; i++)
+    {
+        if (ctx == s_mbedtlsCtx[i])
+        {
+            ret = true;
+            break;
+        }
+    }
+    return ret;
+}
+#endif
+
+#if defined(MBEDTLS_AES_ALT)
+
+/* Implementation that should never be optimized out by the compiler */
+static void mbedtls_zeroize( void *v, size_t n ) {
+    volatile unsigned char *p = (unsigned char*)v; while( n-- ) *p++ = 0;
+}
+
+void mbedtls_aes_init( mbedtls_aes_context *ctx )
+{
+    memset( ctx, 0, sizeof( mbedtls_aes_context ) );
+#if defined(FSL_FEATURE_SOC_LTC_COUNT) && (FSL_FEATURE_SOC_LTC_COUNT > 0)
+    LTC_Init(LTC_INSTANCE);
+#endif
+#if defined(FSL_FEATURE_SOC_DCP_COUNT) && (FSL_FEATURE_SOC_DCP_COUNT > 0)
+    /* Initialize DCP */
+    dcp_config_t dcpConfig;
+
+    DCP_GetDefaultConfig(&dcpConfig);
+    DCP_Init(DCP, &dcpConfig);
+#endif
+}
+
+void mbedtls_aes_free( mbedtls_aes_context *ctx )
+{
+    if( ctx == NULL )
+        return;
+
+    mbedtls_zeroize( ctx, sizeof( mbedtls_aes_context ) );
+#if defined(FSL_FEATURE_SOC_LTC_COUNT) && (FSL_FEATURE_SOC_LTC_COUNT > 0)
+    LTC_Deinit(LTC_INSTANCE);
+#endif
+#if defined(FSL_FEATURE_SOC_DCP_COUNT) && (FSL_FEATURE_SOC_DCP_COUNT > 0)
+    DCP_Deinit(DCP);
+#endif
+}
```

Comment:
Shouldn't there be a crypto_detach_ctx_from_key_slot() in this function as well (at least for some hardware, like MBEDTLS_FREESCALE_CAU3_AES or MBEDTLS_FREESCALE_DCP_AES)?

Reply:
I have added a call to detach for the DCP case.